### ISSUE TITLE
Unable to prevent erroneous encoding on ViewComponents and PartialViews

### ISSUE THREAD

**Issue by CamiloTerevinto** [aspnet/Razor#1375]
The base contextx of this issue can be seen here: [ViewComponent gets wrongly encoded](http://stackoverflow.com/questions/44097576/viewcomponent-gets-wrongly-encoded)

I have been trying for the past couple of hours every combination of these I could think of, both in the View and in the ViewComponent:

 * Html.Raw (string)
 * HtmlRaw (HtmlString)
 * HtmlString
 * string
 * HtmlEncoder = HtmlEncoder.Create(UnicodeRanges.All)
 * HtmlEncoder = HtmlEncoder.Create(TextencoderSettings(UnicodeRanges.All))
 *  Component.InvokeAsync().WriteTo(ViewContext.Writer, HtmlEncoder)
 * Component.InvokeAsync().WriteTo(ViewContextx.Writer, HtmlEncoder.Create(UnicodeRanges.All))
 
And I have no idea how many others I've tried thus far. The output is always the same, even if I look at the InvokeAsync output prior to writing that output to the View. I have even made the Default.cshtml file "unbuildable" to make sure anything was changing.

EDIT: Looking at [this issue ](https://github.com/aspnet/HttpAbstractions/issues/315) and setting the UnicodeRanges.All helps in **Views** (a.k.a. doesn't make me use Html.Raw), but it does nothing on the ViewComponent

EDIT2: The same happens with PartialViews which has never happened to me before

**Comment by CamiloTerevinto** [comment]
I have also tried with 

 * UseRequestLocalization - all Cultures to es
 * CultureInfo.CurrentCulture/CurrentUICulture - to es
* CultureInfol.DefaultThreadCurrentCulture/DefaultThreadCurrentUICulture - to es

Notice that this happens as ViewComponents but I also tried to make one of them a PartialView and there was no change. Only the Index.cshtml renders non latin-basic characters.

**Comment by rynowak** [comment]
> setting the UnicodeRanges.All helps in views

Where are you setting the unicode ranges? Are you replacing the encoder on the view or in startup (`AddWebEncoders`)?

**Comment by CamiloTerevinto** [comment]
I tried through all these ways:

* In the View 
* In the ViewComponent 
* In the PartialView
* Using .AddWebEncoders 
* Using Configure<...Settings> (don't remember the class name)

None changed the output of the ViewComponent/PartialView

**Comment by rynowak** [comment]
That's everything I can think of off the cuff to try. @NTaylorMullen do you have any ideas about this? seems VCTH related?

**Comment by NTaylorMullen** [comment]
@CamiloTerevinto could you provide a running sample? Grabbing bits from other issues/SO I can't seem to reproduce the issue.

**Comment by CamiloTerevinto** [comment]
@NTaylorMullen 
I just tried to create a blank new ASP.NET Core 1.1 project and all (View, PartialView and ViewComponent) render correctly.
Could this be some package (.NET/css/js) in my real solution? I can attach the StartUp class and the Layout if needed

**Comment by NTaylorMullen** [comment]
> Could this be some package (.NET/css/js) in my real solution? I can attach the StartUp class and the Layout if needed

It could. My bet would be on a .NET bit though. If it were js/css you wouldn't be seeing it in the pages source.

**Comment by CamiloTerevinto** [comment]
@NTaylorMullen  in case it makes any difference (no weird package is being used here, however...), here is the Startup:

```
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Identity;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using MySQL.Data.EntityFrameworkCore.Extensions;

    public class Startup
    {
        public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath);

            builder.AddEnvironmentVariables();
            Configuration = builder.Build();
        }

        public IConfigurationRoot Configuration { get; }

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddDbContext<MyClass>(options => options.UseMySQL("CS"));

            services.AddIdentity<User, Role>()
                .AddUserStore<IdentityStore>()
                .AddRoleStore<IdentityStore>()
                .AddDefaultTokenProviders();
            services.Configure<IdentityOptions>(options =>
            {
                options.Cookies.ApplicationCookie.LoginPath = new PathString("/Cuenta/InicioSesion");
                options.Cookies.ApplicationCookie.AccessDeniedPath = new PathString("/Cuenta/InicioSesion");
                options.Cookies.ApplicationCookie.CookieHttpOnly = true;
                options.Cookies.ApplicationCookie.ReturnUrlParameter = "";
                options.Cookies.ApplicationCookie.ExpireTimeSpan = new System.TimeSpan(0, 45, 0);

                options.Password = new PasswordOptions
                {
                    RequireDigit = true,
                    RequiredLength = 8,
                    RequireLowercase = true,
                    RequireUppercase = true
                };
                options.SignIn = new SignInOptions
                {
                    RequireConfirmedEmail = false,
                    RequireConfirmedPhoneNumber = false
                };
            });

            services.AddSingleton<IPasswordHasher<User>, PasswordHasher>();
            services.AddMvc();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseExceptionHandler("/Error");
            }
            
            app.UseDefaultFiles()
                .UseStaticFiles()
                .UseIdentity()
                .UseMvcWithDefaultRoute();

            app.Use((context, next) =>
            {
                context.Response.Headers.Remove("Server");
                return next();
            });
        }
    }
}
```
And this is the Layout:

```
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>@ViewData["Title"]</title>

    <base href="@Url.ActionContext.HttpContext.Request.Scheme://@Url.ActionContext.HttpContext.Request.Host/" />

    <link href="~/css/site.css" rel="stylesheet" media="all" type="text/css" asp-append-version="true" />
    <link href="~/lib/tether/dist/css/tether.min.css" rel="stylesheet" media="all" type="text/css" />
    <link href="~/lib/bootstrap/dist/css/bootstrap.min.css" rel="stylesheet" media="all" type="text/css" />
</head>
<body>
    <nav class="navbar navbar-inverse navbar-toggleable-md bg-inverse" style="padding:0;">
        <div class="container">
            <div class="navbar-header">
                <button type="button" class="navbar-toggler navbar-toggler-right" data-toggle="collapse" data-target=".navbar-collapse">
                    <span class="navbar-toggler-icon"></span>
                </button>
                <a asp-area="" asp-controller="Home" asp-action="Index" class="navbar-brand"></a>
            </div>
            <div class="navbar-collapse collapse">
                @await Html.PartialAsync("_LoginPartial")
            </div>
        </div>
    </nav>
    <div class="container body-content">
        @RenderBody()

        <nav class="navbar fixed-bottom navbar-light bg-faded">
            <p class="text-center" style="margin:0"></p>
        </nav>
    </div>
   
    <script src="~/lib/jquery/dist/jquery.min.js"></script>
    <script src="~/lib/tether/dist/js/tether.min.js"></script>
    <script src="~/lib/bootstrap/dist/js/bootstrap.min.js"></script>

    @RenderSection("Scripts", required: false)
</body>
</html>
```

**Comment by CamiloTerevinto** [comment]
This is the most stupid bug ever.

Since for new projects everything goes just fine, I decided I'd try to create a new project and copy my files. Guess what? **Same problem arose**.
After childishly copying the text to one of the working projects, saving the file and copying the saved file to the project with problems, **it worked just fine**.

I find it completely weird that VS, Notepad and Notepad++ can open the file just fine, but ASP.NET Core can't.

So, solution? Copy the data over and replace the files...

**Comment by yepeekai** [comment]
I had that problem today.  The problem was that for an unknown reason, there was a file using ANSI but files are served or assumed to be UTF8.  I changed the file to UTF8 in notepad++, corrected the incorrect chars and voilà, everything is working ok...  I think visual studio should be more helpful regarding that...  Either give warning or load the correct file Encoding.  This was working ok before we moved to aspnet core 2.
